Function to convert OpenOptions to c_int

FedericoPonzi opened the pull request:
Fixes: #74943
The creation_mode and access_mode function were already available in the OpenOptions struct, but currently private. I've added a new free functions to unix/fs.rs which takes the OpenOptions, and returns the c_int to be used as parameter for the `open` call.

rust-highfive commented:
r? @Mark-Simulacrum

(rust_highfive has picked a reviewer for you, use r? to override)

FedericoPonzi commented:
r? @withoutboats

FedericoPonzi commented:
I would like some guidance on the following:
1. Is that the right place to put the requested function?
2. How can I fix the CI error? Currently failing with `error: function is never used: `get_openopetions_as_cint`. 
3. Were can I put a test for it?

pickfire commented on library/std/src/sys/unix/fs.rs:
```diff
@@ -962,6 +963,12 @@ pub fn rename(old: &Path, new: &Path) -> io::Result<()> {
     Ok(())
 }
 
+pub fn get_openopetions_as_cint(from: OpenOptions) -> io::Result<libc::c_int> {
```
I believe there is a typo here.
```suggestion
pub fn get_openoptions_as_cint(from: OpenOptions) -> io::Result<libc::c_int> {
```

joshtriplett commented on library/std/src/sys/unix/fs.rs:
```diff
@@ -962,6 +963,12 @@ pub fn rename(old: &Path, new: &Path) -> io::Result<()> {
     Ok(())
 }
 
+pub fn get_openoptions_as_cint(from: OpenOptions) -> io::Result<libc::c_int> {
```
This should be an as_flags method on the existing extension trait OpenOptionsExt.

It should operate on &self.

joshtriplett commented on library/std/src/sys/unix/fs.rs:
```diff
@@ -692,6 +692,7 @@ impl OpenOptions {
     }
 }
 
+
```
Please drop this addition.

FedericoPonzi commented:
> I added a couple of comments.
> 
> Also, you should add tests near the existing tests for OpenOptionsExt.

Thanks a lot for your comments! I've added the as_flags method to `OpenOptionsExt` in `unix/ext/fs.rs`. Since the `get_access_mode` and `get_creation_mode` are private to the OpenOptions struct thus I had to duplicate a bit of the code.

I currently have one last issue, I cannot seem to find the location of the tests for `OpenOptionsExt`, could you please tell me where can I find them?

joshtriplett commented:
> Since the `get_access_mode` and `get_creation_mode` are private to the OpenOptions struct thus I had to duplicate a bit of the code.

Like the other methods of `OpenOptionsExt`, you should implement this as a call to `self.inner_mut().as_flags()`, which calls the corresponding method on the `OpenOptions` implemented in at `library/std/src/sys/unix/fs.rs`. That's where `get_access_mode` and `get_creation_mode` are defined, so you'll be able to call them and avoid the duplication.

> I currently have one last issue, I cannot seem to find the location of the tests for `OpenOptionsExt`, could you please tell me where can I find them?

I'd suggest adding doctests to the method, like the other methods of `OpenOptionsExt` have.

pickfire commented on library/std/src/sys/unix/ext/fs.rs:
```diff
@@ -345,6 +345,33 @@ pub trait OpenOptionsExt {
+    /// use std::fs::OpenOptions;
+    /// use std::os::unix::fs::OpenOptionsExt;
+    ///
+    /// # fn main() {
```
I believe `fn main() {` isn't needed here.

FedericoPonzi replied:
I've copied it from the `custom_flags` docs... shall I remove it also from there and from `mode`?

pickfire replied:
We could probably remove that in a follow up patch. Do you want to do it?

FedericoPonzi replied:
sure, I'll do it!

pickfire commented on library/std/src/sys/unix/fs.rs:
```diff
@@ -655,7 +655,11 @@ impl OpenOptions {
+        let access_mode = self.get_access_mode()?;
+        let creation_mode = self.get_creation_mode()?;
+        Ok(creation_mode | access_mode | self.custom_flags)
+    }
```
```suggestion
    }

```
Add extra space here.

pickfire commented on library/std/src/sys/unix/ext/fs.rs:
```diff
@@ -358,6 +385,9 @@ impl OpenOptionsExt for OpenOptions {
         self.as_inner_mut().custom_flags(flags);
         self
     }
```
```suggestion
    }

```
Add extra space here.

pickfire commented on library/std/src/sys/unix/ext/fs.rs:
```diff
@@ -345,6 +345,33 @@ pub trait OpenOptionsExt {
     #[stable(feature = "open_options_ext", since = "1.10.0")]
     fn custom_flags(&mut self, flags: i32) -> &mut Self;
+
+    /// Get the flags of this OpenOptions as libc::c_int.
```
Can we link `libc::c_int`?

FedericoPonzi replied:
Is it enough to put it in brackets like this?
 
    [`libc::c_int`]

FedericoPonzi replied:
Solved using [this](https://github.com/rust-lang/api-guidelines/issues/186) as a reference.

pickfire commented on library/std/src/sys/unix/ext/fs.rs:
````diff
@@ -345,6 +345,33 @@ pub trait OpenOptionsExt {
+    /// # }
+    /// ```
+    #[stable(feature = "open_options_ext_as_flags", since = "1.47.0")]
+    fn as_flags(&self) -> io::Result<libc::c_int>;
````
Is there any reason `libc::c_int` is used instead of `std::os::raw::c_int`?

FedericoPonzi replied:
Just because both `get_access_mode` and `get_creation_mode` use `libc::c_int`. Shall I change the return type to `std::os::raw::c_int` instead?

pickfire replied:
No, I didn't saw the rest.

pickfire commented on library/std/src/sys/unix/ext/fs.rs:
```diff
@@ -345,6 +345,33 @@ pub trait OpenOptionsExt {
+    ///     options.custom_flags(libc::O_NOFOLLOW);
+    /// }
+    /// let file_name = CString::new("foo.txt").unwrap();
+    /// let file = unsafe{ libc::open(file_name.as_c_str().as_ptr(), options.as_flags().unwrap()) };
```
```suggestion
    /// let file = unsafe { libc::open(file_name.as_c_str().as_ptr(), options.as_flags().unwrap()) };
```
Style fix

pickfire commented on library/std/src/sys/unix/ext/fs.rs:
```diff
@@ -345,6 +345,33 @@ pub trait OpenOptionsExt {
+    /// use std::os::unix::fs::OpenOptionsExt;
+    ///
+    /// # fn main() {
+    /// let mut options = OpenOptions::new();
+    /// options.write(true).read(true);
```
```suggestion
    /// let options = OpenOptions::new().write(true).read(true);
```
We could just reuse `mut`.

FedericoPonzi replied:
I've encountered the following error:
```
error[E0716]: temporary value dropped while borrowed
  --> src/sys/unix/ext/fs.rs:362:19
   |
10 | let mut options = OpenOptions::new().write(true).read(true);
   |                   ^^^^^^^^^^^^^^^^^^                       - temporary value is freed at the end of this statement
   |                   |
   |                   creates a temporary which is freed while still in use
11 | let file_name = CString::new("foo.txt").unwrap();
12 | let file = unsafe { libc::open(file_name.as_c_str().as_ptr(), options.as_flags().unwrap()) };
   |                                                               ------- borrow later used here
   |
   = note: consider using a `let` binding to create a longer lived value

```
Thus I would keep the old version, which is also similar to what can be found on the other function's docs if that's ok

pickfire commented on library/std/src/sys/unix/fs.rs:
```diff
@@ -655,6 +655,11 @@ impl OpenOptions {
     pub fn mode(&mut self, mode: u32) {
         self.mode = mode as mode_t;
     }
```
Missing space
```suggestion
    }

```

FedericoPonzi replied:
Thanks Ivan for all your good comments!

pickfire commented on library/std/src/sys/unix/ext/fs.rs:
```diff
@@ -345,6 +345,31 @@ pub trait OpenOptionsExt {
     fn custom_flags(&mut self, flags: i32) -> &mut Self;
+
+    /// Get the flags of this OpenOptions as [`libc::c_int`].
+    /// With: [`libc::open`]
```
> With: [`libc::open`]

What does this mean?

pickfire commented on library/std/src/sys/unix/ext/fs.rs:
```diff
@@ -345,6 +345,31 @@ pub trait OpenOptionsExt {
+    /// Get the flags of this OpenOptions as [`libc::c_int`].
+    /// With: [`libc::open`]
+    ///
+    /// This method allows the reuse of the OpenOptions as flags argument for [`fs::OpenOptions`].
```
I also don't quite understand this.

pickfire commented on library/std/src/sys/unix/fs.rs:
```diff
@@ -655,6 +655,11 @@ impl OpenOptions {
+    pub fn as_flags(&self) -> io::Result<c_int> {
+        let access_mode = self.get_access_mode()?;
+        let creation_mode = self.get_creation_mode()?;
+        Ok(creation_mode | access_mode | self.custom_flags)
```
Why not just make use of `self.mode`?

FedericoPonzi replied:
because  If you say .read(true).write(true). is self.mode reflecting these changes? I doesn't look like it from the code

pickfire commented on library/std/src/sys/unix/ext/fs.rs:
```diff
@@ -345,6 +345,31 @@ pub trait OpenOptionsExt {
     #[stable(feature = "open_options_ext", since = "1.10.0")]
     fn custom_flags(&mut self, flags: i32) -> &mut Self;
+
+    /// Get the flags of this OpenOptions as [`libc::c_int`].
```
```suggestion
    /// Get the flags as [`libc::c_int`].
```
`of this OpenOptions` seemed unnecessary

FedericoPonzi commented:
r? @joshtriplett 
I've r you because you already approved the PR, and this hadn't any traffic for a while now and I'm not sure how to move this forward. Thanks!

jyn514 commented on library/std/src/sys/unix/ext/fs.rs:
````diff
@@ -345,6 +345,30 @@ pub trait OpenOptionsExt {
+    /// let file_name = CString::new("foo.txt").unwrap();
+    /// let file = unsafe { libc::open(file_name.as_c_str().as_ptr(), options.as_flags().unwrap()) };
+    /// ```
+    #[stable(feature = "open_options_ext_as_flags", since = "1.47.0")]
````
I don't think this should be insta-stable. Usually new libs features go through at least a few cycles of nightly.
```suggestion
    #[unstable(feature = "open_options_ext_as_flags", issue = "none")]
```
Make sure to open a tracking issue and replace "none" with the issue number.

jyn514 commented:
@bors r=@JoshTriplet

bors commented:
:pushpin: Commit 2f5192280f57ab859a2414ff2e9b6f6398d3feb7 has been approved by `JoshTriplet`

<!-- @bors r=JoshTriplet 2f5192280f57ab859a2414ff2e9b6f6398d3feb7 -->
<!-- homu: {"type":"Approved","sha":"2f5192280f57ab859a2414ff2e9b6f6398d3feb7","approver":"JoshTriplet"} -->

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #76626

bors commented:
:pushpin: Commit 2f5192280f57ab859a2414ff2e9b6f6398d3feb7 has been approved by `JoshTriplett`

<!-- @bors r=JoshTriplett 2f5192280f57ab859a2414ff2e9b6f6398d3feb7 -->
<!-- homu: {"type":"Approved","sha":"2f5192280f57ab859a2414ff2e9b6f6398d3feb7","approver":"JoshTriplett"} -->

bors commented:
:hourglass: Testing commit 2f5192280f57ab859a2414ff2e9b6f6398d3feb7 with merge e0bc267512fc0cb49c86978192857e8187017f0b...
<!-- homu: {"type":"BuildStarted","head_sha":"2f5192280f57ab859a2414ff2e9b6f6398d3feb7","merge_sha":"e0bc267512fc0cb49c86978192857e8187017f0b"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1150184282), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36777)
Approved by: JoshTriplett
Pushing e0bc267512fc0cb49c86978192857e8187017f0b to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"JoshTriplett","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36777","checks-actions":"https://github.com/rust-lang-ci/rust/runs/1150184282"},"merge_sha":"e0bc267512fc0cb49c86978192857e8187017f0b"} -->
